### PR TITLE
TypeScript update!!!

### PR DESCRIPTION
**Pull request by ofluffydev**
# Pull Request

## Description:

Added full typescript, prettier, and setup jest for unit testing. (Next part will probably be writing those tests)

Also added a README with details on developing/running the node

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive `README.md` file detailing the Purple Bubble Server's features, setup instructions, and API endpoints.
	- Added a job scheduling mechanism to synchronize server data and messages at specified intervals.
	- Implemented a basic Express server for message handling and synchronization with Redis.
	- Created a `.prettierignore` file to specify files to be ignored by the Prettier code formatter.
	- Established a `tsconfig.json` file for TypeScript compiler options.

- **Improvements**
	- Created configuration files for Prettier and Jest to enhance code formatting and testing capabilities.
	- Added TypeScript interfaces for better type safety in server and message data handling.
	- Enhanced formatting and readability of various documentation files, including `CODE_OF_CONDUCT.md` and `CONTRIBUTING.md`.

- **Bug Fixes**
	- Minor formatting adjustments across several documentation files for improved readability.

- **Chores**
	- Updated project dependencies and scripts in `package.json` for better development workflow.

### CODE REVIEW

File: src/index.ts
Comment (jaspermayone) [reply src/index.ts]: @ofluffydev pls do ths!

Comment (jaspermayone) [reply src/index.ts]: *this

**Review by jaspermayone — CHANGES_REQUESTED** [review]
Pls see Coderabit suggestions then rerequest my review.

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]: I think we should probably remove this line.

Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion

```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md:8]:
```suggestion

```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md:51]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md:50]:
```suggestion
> ### Legal Notice
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (jaspermayone) [inline CONTRIBUTING.md]:
```suggestion
```

File: LICENSE.md
Comment (jaspermayone) [inline LICENSE.md:223]:
```suggestion
- a) The work must carry prominent notices stating that you modified
  it, and giving a relevant date.
  
- b) The work must carry prominent notices stating that it is
  released under this License and any conditions added under
  section 7. This requirement modifies the requirement in section 4
  to "keep intact all notices".
  
- c) You must license the entire work, as a whole, under this
  License to anyone who comes into possession of a copy. This
  License will therefore apply, along with any applicable section 7
  additional terms, to the whole of the work, and all its parts,
  regardless of how they are packaged. This License gives no
  permission to license the work in any other way, but it does not
  invalidate such permission if you have separately received it.
  
- d) If the work has interactive user interfaces, each must display
  Appropriate Legal Notices; however, if the Program has interactive
  interfaces that do not display Appropriate Legal Notices, your
  work need not make them do so.
```

File: README.md
Comment (jaspermayone) [inline README.md:17]: We should consider bumping this to a much more current version of node, and pinning the dependency in the package file.

(See https://github.com/phish-directory/api/blob/4eb090625fb7b03251e7336310f12a255845affb/package.json#L60-L64 as an example)

File: README.md
Comment (jaspermayone) [inline README.md]:
```suggestion
pnpm run build
```

File: README.md
Comment (jaspermayone) [inline README.md]:
```suggestion
## Installation

1. Clone the repository:

   ```bash
   git clone git@github.com:thepurplebubble/server-node.git
   ```

2. Navigate to the project directory:

   ```bash
   cd server-node
   ```

3. Install dependencies:

   ```bash
   pnpm install
   ```

4. Create a `.env` file in the root directory with the following content:
   ```env
   PORT=8080
   REDIS_URL=redis://localhost:6379
   ```
   Adjust these values as needed for your environment.

```

File: README.md
Comment (jaspermayone) [inline README.md]:
```suggestion
```bash
```

File: README.md
Comment (jaspermayone) [inline README.md]:
```suggestion
```bash
```

File: README.md
Comment (jaspermayone) [inline README.md:60]: @ofluffydev is this still a todo? If so can we move it to an issue instead of an inline todo as this is cleaner.

File: README.md
Comment (jaspermayone) [inline README.md]: Replace email with the team email.

```suggestion
For other inquiries, please reach out to [team@purplebubble.org](mailto:team@purplebubble.org).
```

File: README.md
Comment (jaspermayone) [inline README.md]:
```suggestion
```bash
pnpm run dev
```
```

File: package.json
Comment (jaspermayone) [inline package.json]:
```suggestion
  "main": "src/index.ts",
```

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts]:
```suggestion
import express from "express";
import bodyParser from "body-parser";
import { createClient } from "redis";
import * as dotenv from 'dotenv'; 
dotenv.config();
```

File: src/index.ts
Comment (jaspermayone) [reply src/index.ts]: @ofluffydev this stands pls move to separate file!

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts:29]: Convert this to an issue instead of inline for better tracking and contributions plz!

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts]: Not standard at all.
```suggestion
  res.status(200).send("ok");
```

File: src/util.ts
Comment (jaspermayone) [inline src/util.ts:83]: Pls address todo or make a github issue.

File: yarn.lock
Comment (jaspermayone) [inline yarn.lock:1]: @ofluffydev let's use PNPM. Also, I think lockfiles aren’t supposed to be commited, I could be wrong though.

**Review by jaspermayone — CHANGES_REQUESTED** [review]
Overall super good! Very exited to have this project still going, tysm! 

Lmk if you have any questions on my comments.

File: src/index.ts
Comment (ofluffydev) [reply src/index.ts:29]: I didn't make that comment 💀

File: src/util.ts
Comment (ofluffydev) [reply src/util.ts:83]: also didn't write this comment. this update is focused on typescript

File: yarn.lock
Comment (ofluffydev) [reply yarn.lock:1]: I thought the bot auto removed it but okay

**Comment by ofluffydev** [comment]
I'll look at the import and environment variables issues but I cannot tackle every single todo comment at once please 😭

None of those comments were added by me

File: src/util.ts
Comment (jaspermayone) [reply src/util.ts:83]: Ok. Can you create a github issue for this then

**Comment by jaspermayone** [comment]
We can convert the todo comments to issues, even if not added by you.

File: README.md
Comment (ofluffydev) [reply README.md:60]: Yeah I think later I mentioned making docs with docusaurus instead and linking them there. I'll open issue

File: README.md
Comment (ofluffydev) [reply README.md:17]: Yeah I think 18 is the newest LTS now, not sure if we wanna go for LTS or latest?

**Comment by ofluffydev** [comment]
Merge conflicts caused by package files, I think I have commit permissions to push this PR, but didn't want to without you re-reviewing like you asked.

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts]: You don't nee to send the error code again!

```suggestion
      res.status(400).send("Bad Request");
```

**Comment by jaspermayone** [comment]
Am working on reviewing @ofluffydev!

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts]: this is a JS file?

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts:4]: this is a file? or? configs should probably be stored in a JSON file if not sensitive or .env for sensitive

File: src/jobs.ts
Comment (jaspermayone) [inline src/jobs.ts]: if these are all TS things, you can leave off the file extension and it'll compile with the right extension

File: src/jobs.ts
Comment (jaspermayone) [reply src/jobs.ts:41]: @ofluffydev

**Review by jaspermayone — CHANGES_REQUESTED** [review]

File: src/index.ts
Comment (ofluffydev) [reply src/index.ts]: what?

File: src/index.ts
Comment (ofluffydev) [reply src/index.ts:4]: No it's just stuff that runs on startup

File: src/jobs.ts
Comment (ofluffydev) [reply src/jobs.ts]: which requires one rule to be changed which requires more rules to be changed and then I think it eventually needs an entire different ES version

File: src/jobs.ts
Comment (ofluffydev) [reply src/jobs.ts:41]: Really don't know what code rabbit is on about or what redis scan does it's just changed to work with typescript

**Comment by ofluffydev** [comment]
there was an issue with importing ts with the current setup I forgot what it was. something related to this config option rich requires even more options to be changed.

https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts]:
```suggestion
import { scheduleJobs } from "./jobs";
```

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts:55]: This is redundant, 200 is the code for ok

```suggestion
  res.status(200);
```

File: src/index.ts
Comment (jaspermayone) [inline src/index.ts:42]:
```suggestion
      res.status(400);
```
400 is the error code, unless you are providing more info, having bad request there is redundant

**Review by jaspermayone — CHANGES_REQUESTED** [review]

**Comment by jaspermayone** [comment]
> there was an issue with importing ts with the current setup I forgot what it was. something related to this config option rich requires even more options to be changed.
> 
> https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions

@ofluffydev you leave off the file extension and it works.

**Comment by jaspermayone** [comment]
(without changing any settings)

File: src/jobs.ts
Comment (jaspermayone) [reply src/jobs.ts]: No, you can take the file names off and it works just fine. I've used it before in ts projects.

Comment (jaspermayone) [reply src/jobs.ts]:
```suggestion
import { syncServers, storeMessage } from "./util";
import { Server, Message } from "./types";
```

File: src/index.ts
Comment (ofluffydev) [reply src/index.ts:42]: It's not redundant that's the how the request gets sent, only setting the status, well only sets the status 😭

Comment (ofluffydev) [reply src/index.ts:42]: Actual solution would be to make the text more descriptive, maybe even send more details with json. Not sending anything is worse than sending bad request .

File: src/index.ts
Comment (ofluffydev) [reply src/index.ts:55]: I think we wanna send the actual response so...

**Comment by ofluffydev** [comment]
again, this is meant to be a typescript update not fixing every issue the existing code had. can we please get this merged and if you don't like the response bodies or how logic is being handled then we can open issues.

**Comment by jaspermayone** [comment]
Code should be committed to prod without glaring holes, I will do another review though, as I think it should be pretty good now!

**Comment by jaspermayone** [comment]
@coderabbitai full review

File: src/jobs.ts
Comment (jaspermayone) [inline src/jobs.ts]:
```suggestion
import { redis } from "./index";
```

**Review by jaspermayone — DISMISSED** [review]

**Pull request merged**
